### PR TITLE
remove for-loops

### LINKED ISSUE #3861: Why is a for-loop used here?
At <https://github.com/Microsoft/TypeScript/blob/master/src/services/services.ts#L272-279>?

Could it be replaced by
```ts
            if (children.length === 0) { return undefined; }

            let child = children[children.length - 1];

            if (child.kind < SyntaxKind.FirstNode) {
                return child;
            }

            return child.getLastToken(sourceFile);
```
with a similar modification for `getFirstToken()` above?

### PR DESCRIPTION
**Pull request by aozgaa**
Fixes #3861

### CODE REVIEW

File: src/services/services.ts [hunk 1]
```diff
@@ -260,26 +260,30 @@ namespace ts {
         }
 
         public getFirstToken(sourceFile?: SourceFile): Node {
-            let children = this.getChildren();
-            for (let child of children) {
-                if (child.kind < SyntaxKind.FirstNode) {
-                    return child;
-                }
+            let children = this.getChildren(); // why isn't sourceFile passed as an argument??
+            if (!(children && children.length > 0)) { return undefined; }
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Put the body on a different line

File: src/services/services.ts [hunk 2]
```diff
@@ -260,26 +260,30 @@ namespace ts {
...
         }
 
         public getLastToken(sourceFile?: SourceFile): Node {
             let children = this.getChildren(sourceFile);
-            for (let i = children.length - 1; i >= 0; i--) {
-                let child = children[i];
-                if (child.kind < SyntaxKind.FirstNode) {
-                    return child;
-                }
+            if (!children) { return undefined; }
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Body on different line

File: src/services/services.ts [hunk 3]
```diff
@@ -260,26 +260,30 @@ namespace ts {
...
+
+            let child = lastOrUndefined(children);
+            if (!child) { return undefined; }
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Body on different line

File: src/services/services.ts (hunk 2 above)
Comment (DanielRosenwasser) [inline src/services/services.ts]: `children` always seems to be defined, just make this `children.length === 0`

File: src/services/services.ts (hunk 1 above)
Comment (DanielRosenwasser) [inline src/services/services.ts]: Same here regarding children

Comment (DanielRosenwasser) [inline src/services/services.ts]: You should pass `sourceFile in as an argument

File: src/services/services.ts [hunk 4]
```diff
@@ -260,26 +260,30 @@ namespace ts {
         }
 
         public getFirstToken(sourceFile?: SourceFile): Node {
-            let children = this.getChildren();
-            for (let child of children) {
-                if (child.kind < SyntaxKind.FirstNode) {
-                    return child;
-                }
+            let children = this.getChildren(); // why isn't sourceFile passed as an argument??
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: You should pass `sourceFile` in as an argument

File: src/services/services.ts (hunk 2 above)
Comment (aozgaa) [inline src/services/services.ts]: `getLastOrUndefined()` checks whether the array is empty and returns the last value or undefined accordingly.

File: src/services/services.ts (hunk 1 above)
Comment (aozgaa) [inline src/services/services.ts]: changed.

File: src/services/services.ts (hunk 3 above)
Comment (aozgaa) [inline src/services/services.ts]: Changed.

File: src/services/services.ts (hunk 4 above)
Comment (aozgaa) [inline src/services/services.ts]: Changed

File: src/services/services.ts [hunk 5]
```diff
@@ -260,26 +260,25 @@ namespace ts {
...
         public getFirstToken(sourceFile?: SourceFile): Node {
-            let children = this.getChildren();
-            for (let child of children) {
-                if (child.kind < SyntaxKind.FirstNode) {
-                    return child;
-                }
-
-                return child.getFirstToken(sourceFile);
+            let children = this.getChildren(sourceFile);
+            if (!(children.length > 0)) {
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Either `if (!children.length)` or `if (children.length === 0)`

Comment (aozgaa) [inline src/services/services.ts]: Changed :)

**Comment by DanielRosenwasser** [comment]
:+1:

**Pull request merged**